App Inspector (#539)
localstack/localstack-docs: src/content/docs/aws/capabilities/web-app/app-inspector.md

---
title: App Inspector
description: App Inspector is a LocalStack tool designed to enhance system observability by enabling users to view, collect, and inspect data exchanges, including event payloads and metadata, between AWS services.
template: doc
tags: ["Hobby"]
---

## Introduction

App Inspector allows users to view, collect, and inspect data exchanges, including event payloads and metadata, between AWS services.
It enhances system observability by displaying the data exchanged at every stage, facilitating clear understanding of event flows.

> It further centralizes debugging by enabling event replay at any system stage for detailed analysis and troubleshooting.

Reviewer: Replay won't be supported in the first release.

In addition, it serves as a single point of truth to understand potential errors, service configuration mismatches and IAM permission issues.

With App Inspector, you can:

> - Observe and understand the flow of events through your system.

Reviewer: Let's stay away from using "event", but instead say "flow of operations". We found that "event" is too heavily associated with event-driven development, and was making non-EDD people feel it wasn't for them.

- Identify errors and obtain detailed information for corrections.
- Get immediate feedback on any misconfigurations in your services.
- Gain insights into IAM policies and detect missing permissions.

> - Enhance event contexts for advanced flow tracing.

Reviewer: I'm not sure what this bullet point means. I suggest removing it, unless you can use simpler language here.

## Use Cases

> **Debugging event routing failures**

Reviewer: This whole sub-section needs to be rephrased. It should be focused on "flow of information between services", which is a key selling point of App Inspector.

The current explanation is too focused on events / EventBridge, which makes it feels like a point solution for events, rather than a general solution for understanding flow of information.

When an event doesn't reach its intended target, App Inspector shows exactly where it stopped — a mismatched EventBridge rule pattern, an incorrect ARN, or a missing SQS event source mapping. You get a precise answer without deploying to AWS or adding instrumentation.

**Verifying service-to-service payloads**

> Inspect the exact data passed between services at every hop of your workflow. This is especially useful for confirming that a Lambda is passing the right payload to SNS, or that an SQS message body matches what the downstream consumer expects — before it becomes a production issue.

Reviewer: Instead of "This is especially useful", I'd suggest "For example".

That is, the Lambda -> SNS example is just one use case, rather than a special case.

**Catching IAM misconfigurations early**

> Mocks return whatever you tell them to return, so they can't surface permission errors. App Inspector runs against the real LocalStack emulator, which enforces IAM policies. Missing or overly restrictive permissions show up immediately, at the point of failure in the event chain.

Reviewer: This sentence needs to focus on "finding IAM permission problems in the emulator, before deployment to the cloud where they're much harder to find".

Also, I'm not sure why "Mocks" is mentioned here. That's not relevant at all.

## LocalStack Toolkit for VS Code

App Inspector is built into the [LocalStack Toolkit for VS Code](/aws/tooling/vscode-extension), so you can trace event flows and inspect payloads without leaving your editor. Once LocalStack is running, open the App Inspector panel directly in LocalStack Toolkit extension to view events, and drill into service interactions.

If you haven't set up the toolkit yet, see [LocalStack Toolkit for VS Code](/aws/tooling/vscode-extension) to get started.

![App Inspector panel in the LocalStack Toolkit for VS Code showing a live event stream](/images/aws/app-inspector-vscode.png)

> ## Web Application

Reviewer: Should we be calling this "LocalStack Console"? I'm not sure how consistently that term is used in the docs.

Reviewer 2: in the docs we consistently use the term web app, so this is correct + consistent

You can access App Inspector directly in the [LocalStack Web Application](https://app.localstack.cloud/inst/default/appinspector/spans). Navigate to **App Inspector** in your browser to view events captured from your running LocalStack instance, inspect payloads, and trace service connections.

### Features

### List the local events

> With Event Studio, you can view a detailed list of events in your application, including event producers, types, and timestamps.

Reviewer: "Event Studio" shouldn't be mentioned here. Also, please use "Operations" instead of "Events" in all writing.

The interface enables you to trace the flow of events, identify relationships between services, and analyze patterns for debugging or optimization.

> 

Reviewer: I feel we need a lot more detail here. In particular:

1) What each field on the table represents (e.g. Producer, Consumer, Action, etc).
2) The fact that some operations have errors associated with them.
3) How you can click on the "View Graph" to switch to a graph view.

Reviewer 2: @peter-smith-phd @sheltongraves has addressed your feedback via this commmit:
https://github.com/localstack/localstack-docs/pull/539/changes/8e90021d802f118d00604d89d7f6d73916774ccd 

resolving comment 👌🏻

![App Inspector event list showing a fanout pipeline across EventBridge, Lambda, SQS, and SNS](/images/aws/app-inspector-fanout-pipeline.png)

> ### View Event Details

Reviewer: Before talking about the "View Operation Details" page, we should talk about the Graph view, and how it can be navigated.

Alternatively, rename this section "The Graph View" and merge the content for both the Graph and the Operation Detail panel into a single narrative.

By clicking on an events details, you can visualize the relationships between AWS services in your event-driven architecture and drill down into the specifics of each interaction.
App Inspector displays an interactive graph showing how services connect, alongside detailed payloads, metadata, and status for each operation.
It also highlights errors, warnings, and potential IAM permission issues, enabling precise debugging and troubleshooting.

![App Inspector operation details panel showing service, resource, payloads, and status for a Lambda invocation](/images/aws/app-inspector-operation-details.png)

## Plans & Tiers

| Feature | Hobby | Base / Ultimate / Enterprise |
|---|---|---|

> | Maximum events stored | 10 | 1,000 |

Reviewer: Please change "events" to "operations"

We also need a discussion about selecting "Clear Operations" to make room for more operations to be stored.

| Error analysis | — | ✓ |
| Request payload | ✓ | ✓ |
| Response payload | — | ✓ |

## Supported Services

> The following services are supported on App Inspector:

Reviewer: This is where we need more detail on what "supported" actually means. In theory, _all_ operations for _all_ services should appear, but the services in this list have been optimized to improve their visual layout.

> - [S3](/aws/services/s3)

Reviewer: API Gateway V1 should be on this list.

API Gateway V2 will be added, assuming we get it done in time.

The interface enables you to trace the flow of operations, identify relationships between services, and analyze patterns for debugging or optimization.

![App Inspector event list showing a fanout pipeline across EventBridge, Lambda, SQS, and SNS](/src/assets/images/aws/app-inspector/app-inspector-fanout-pipeline.png)

### Graph View and Operation Details

Click **View Graph** to visualize the relationships between AWS services in your application as an interactive graph. Each node represents a service, and each edge represents a call between them, making it easy to follow the path of a request across your architecture.

By clicking on an operation in the graph or list, you can drill down into the specifics of each interaction. App Inspector displays detailed payloads, metadata, and status for each operation, and highlights errors, warnings, and potential IAM permission issues, enabling precise debugging and troubleshooting.

![App Inspector operation details panel showing service, resource, payloads, and status for a Lambda invocation](/src/assets/images/aws/app-inspector/app-inspector-operation-details.png)

## Supported Services

> ```suggestion

Reviewer 2: ```suggestion

```

---
title: App Inspector
description: App Inspector is a LocalStack tool designed to enhance system observability by enabling users to view, collect, and inspect data exchanges, including event payloads and metadata, between AWS services.
template: doc
tags: ["Hobby"]
---

## Introduction

App Inspector allows users to view, collect, and inspect data exchanges, including payloads and metadata, between AWS services.
It enhances system observability by displaying the data exchanged at every stage, facilitating clear understanding of operation flows.
In addition, it serves as a single point of truth to understand potential errors, service configuration mismatches, and IAM permission issues.

With App Inspector, you can:

- Observe and understand the flow of operations through your system.
- Identify errors and obtain detailed information for corrections.
- Get immediate feedback on any misconfigurations in your services.
- Gain insights into IAM policies and detect missing permissions.

## Requirements

> App Inspector requires **LocalStack 2026.03.0** or later.

Reviewer: ```suggestion
App Inspector requires **LocalStack 2026.04.0** or later.
```

## Use Cases

**Understanding service-to-service information flow**

When a service call doesn't reach its intended target, App Inspector shows exactly where it stopped and why — tracing the flow of information between services at every hop. You get a precise answer without deploying to AWS or adding instrumentation.

**Verifying service-to-service payloads**

> Inspect the exact data passed between services at every hop of your workflow. For example, confirming that a Lambda is passing the right payload to SNS, or that an SQS message body matches what the downstream consumer expects — before it becomes a production issue.

Reviewer: ```suggestion
Inspect the exact data passed between services at every hop of your workflow. For example, confirming that a Lambda function is passing the correct payload to SNS, or that an SQS message body matches what the downstream consumer expects — before it becomes a production issue.
```

**Catching IAM misconfigurations early**

> App Inspector runs against the real LocalStack emulator, which enforces IAM policies. Missing or overly restrictive permissions show up immediately — in your local environment, where they're fast and cheap to fix, rather than at deployment time in the cloud.

Reviewer: ```suggestion
App Inspector runs against the LocalStack emulator, which monitors use of IAM policies. Missing or overly restrictive permissions show up immediately — in your local environment, where they're fast and cheap to fix, rather than at deployment time in the cloud.
```

:::note
Operations generated during resource creation will appear in App Inspector alongside your application operations. To keep your trace clean, click **Clear Operations** before sending requests so only the relevant activity is captured.
:::

### Features

### List Operations

You can view a detailed list of operations in your application, including the **Producer** (the service that initiated the call), **Consumer** (the target service), **Action** (the specific API call made), and **Timestamp**. Operations that encountered errors are flagged inline, so you can identify failures at a glance. Click **View Graph** on any operation to switch to the graph view and see how services connect.

The interface enables you to trace the flow of operations, identify relationships between services, and analyze patterns for debugging or optimization.

![App Inspector event list showing a fanout pipeline across EventBridge, Lambda, SQS, and SNS](/src/assets/images/aws/app-inspector/app-inspector-fanout-pipeline.png)

### Graph View and Operation Details

> Click **View Graph** to visualize the relationships between AWS services in your application as an interactive graph. Each node represents a service, and each edge represents a call between them, making it easy to follow the path of a request across your architecture.

Reviewer: The latest behaviour is that you just click anywhere on the row, and it'll switch to the graph view. There is still a `View Details` button that is part of that row, but that has the same effect. There is nothing that says `View Graph` (although we did talk about that option).

By clicking on an operation in the graph or list, you can drill down into the specifics of each interaction. App Inspector displays detailed payloads, metadata, and status for each operation, and highlights errors, warnings, and potential IAM permission issues, enabling precise debugging and troubleshooting.

![App Inspector operation details panel showing service, resource, payloads, and status for a Lambda invocation](/src/assets/images/aws/app-inspector/app-inspector-operation-details.png)

## Plans & Tiers

| Feature | Hobby | Base / Ultimate / Enterprise |
|---|---|---|
| Maximum operations stored | 10 | 1,000 |
| Error analysis | — | ✓ |
| Request payload | ✓ | ✓ |
| Response payload | — | ✓ |

## Supported Services

> 

Reviewer 2: [We're going to consistently store all Licensing & Tiers table plans in same doc](https://github.com/localstack/localstack-docs/pull/616), singular source of truth.

```suggestion
## Supported Services

```suggestion

:::note
Visit our [Licensing & Tiers](https://docs.localstack.cloud/aws/licensing/) doc to explore usage with App Inspector.

:::

```
